src/lib.rs

```diff
@@ -315,59 +321,90 @@ pub fn run(config: Config) -> Result<(), Box<dyn Error>> {
     Ok(())
 }
 
-#[derive(Debug, PartialEq)]
-enum CliCommand {
-    Start,
-}
+pub fn parse_args<I, T>(args: I) -> CliArgs
+where
+    I: IntoIterator<Item = T>,
+    T: Into<OsString> + Clone,
+{
+    let app_matches = App::new(clap::crate_name!())
+        .version(clap::crate_version!())
+        .author(clap::crate_authors!())
+        .about(clap::crate_description!())
+        .setting(AppSettings::SubcommandRequiredElseHelp)
+        .subcommand(
+            SubCommand::with_name("start")
+                .about("Start a new tmux session")
+                .arg(
+                    Arg::with_name("PROJECT_FILE")
+                        .help("The session project file to launch")
+                        .required(true),
+                ),
+        )
+        .get_matches_from(args);
+
+    let (command_name, command_matches) = match app_matches.subcommand() {
+        (name, Some(matches)) => (name, matches),
+        (_, None) => {
+            panic!("Subcommand should be forced by clap");
+        }
+    };
 
-impl CliCommand {
-    fn new(maybe_command: &str) -> Result<CliCommand, String> {
-        match maybe_command {
-            "start" => Ok(Self::Start),
-            // TODO: present static list of valid options instead?
-            _ => Err(format!("Command ({}) not recognized.", maybe_command)),
+    let command = match CliCommand::from_str(command_name) {
+        Ok(command) => command,
+        Err(error) => {
+            panic!(error.to_string());
         }
+    };
+
+    let project_name = command_matches
+        .value_of("PROJECT_FILE")
+        .expect("project file is required by clap")
+        .to_string();
+
+    CliArgs {
+        command,
+        project_name,
     }
 }
 
 #[derive(Debug, PartialEq)]
-pub struct CliArgs {
-    command: CliCommand,
-    project_name: String,
+pub enum CliCommand {
+    Start,
 }
 
-impl CliArgs {
-    pub fn new(args: &[String]) -> Result<CliArgs, String> {
-        // TODO: None of this scales very well, but I wanted to see if I could
-        // avoid using a third-party library. Maybe it's worth trying clap or
-        // quicli.
+#[derive(Debug)]
+pub struct ParseCliCommandError;
 
-        let args_ = args.to_owned();
-        // drop entrypoint (e.g. ./rmuxinator)
-        let mut args_ = args_.iter().skip(1);
-        let command_ = args_.next();
-        let project_ = args_.next();
-
-        if command_.is_none() {
-            return Err(String::from("Command is required."));
-        }
+// TODO: this boilerplate can be cut down by using a third-party crate
```

Comment:
Which crate did you have in mind? I'm not opposed.

Reply:
There are a few good options here because this is an application instead of a library. If it was a library I would suggest just [`thiserror`](https://lib.rs/crates/thiserror) where the config error can be represented by something like

```rust
#[derive(thiserror::Error, Debug)]
pub enum ConfigError {
    // TODO: This could be divided up based on the IO error (permission denied, missing, etc.)
    #[error("Unable to open the provided project file")]
    ProjectFileError,
    #[error("Unable to read project file")]
    ProjectReadError,
    // Just expose the underlying toml error
    #[error(transparent)]
    DeserializationError(#[from] toml::de::Error),
}
```

This is recommended for libraries because it's just providing a convenient way to set up using normal old `std::error::Error` instead of forcing people downstream to pull in third-party crates for custom error handling too.

This works just fine for applications too, but there's the change of where errors are often bubbled up to the top level in applications whereas libraries are fine to expose several different error types. Bubbling all these errors up to the same place normally requires some `Box<dyn Error>` trickery, or implementing `From` to convert between errors. So for this program you might have some `ConfigError`, `CliError`, and some other errors that would all need to be converted to some common error to be returned from `main`. This adds in some more boilerplate which can be cut down by using something like [`anyhow`](https://lib.rs/crates/anyhow) in conjunction with `thiserror` (they were made by the same person as well). `anyhow` serves the purpose of providing a nice way to return any kind of error that implements `std::error::Error` and provides a nicely formatted message when you return the error from `main`.

The other option I normally hear about for applications is [`snafu`](https://lib.rs/crates/snafu) which provides very similar functionality to `thiserror` + `anyhow`.

This is probably a bit longer of a response than you were expecting, but "proper" error handling is on my list of stuff that would be good to work through, so I've been judging what all needs to be changed passively :smile:

Reply:
That all makes sense. Maybe we kick the can on this?

Reply:
I'd agree, this should really all be handled in a separate PR. I'll open up an issue going over how all the different options look when I get time.